ldk/csharp/OliveHelpsLDK/OliveHelpsLDK/Filesystem/File.cs

```diff
@@ -0,0 +1,286 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Google.Protobuf;
+using Grpc.Core;
+using OliveHelpsLDK.Logging;
+using Proto;
+
+namespace OliveHelpsLDK.Filesystem
+{
+    internal enum FileStatus
+    {
+        Pending,
+        Initialized,
+        Closed,
+        Errored
+    }
+
+    /// <summary>
+    /// Implementation of <see cref="IFile"/>.
+    /// </summary>
+    /// <remarks>
+    /// We only know if there's a failure when a request is attempted and fails.
+    /// https://github.com/grpc/grpc/issues/25101 has been opened to add this capability in a more intuitive fashion.
+    /// </remarks>
+    internal class File : IFile
+    {
+        internal AsyncDuplexStreamingCall<FilesystemFileStreamRequest, FilesystemFileStreamResponse> Stream { get; }
+        private Session Session { get; }
+
+        private ILogger Logger { get; set; }
+
+        private string Path { get; set; }
+
+        private FileStatus Status { get; set; }
+
+        private TaskCompletionSource<bool> CompletionSource { get; }
+
+        internal File(AsyncDuplexStreamingCall<FilesystemFileStreamRequest, FilesystemFileStreamResponse> stream,
```

Comment:
I'll need to put in a interceptor for this type of call too, so it's fine if this goes in first